### PR TITLE
RFC pyodide governance and decision-making process

### PR DESCRIPTION
**Pull request by rth**
This is an initial draft and request for comments for the governance document aiming to formalize the decision making process for pyodide, done as part of the migration to a separate org (https://github.com/iodide-project/pyodide/issues/1217) 

In practice this is useful to have a transparent process for approving new maintainers, accepting contributions and making other project related decisions. This draft is a close adaptation and simplification of the scikit-learn [governance document](https://scikit-learn.org/stable/governance.html). I think it's more or less how we and other scientific python projects function. I tried to keep it simple, but it's still somewhat detailed.

*TLDR*: the general approach is to try to reach consensus, and if a consensus cannot be reached call for a vote among core developers.

Alternatively, this document could also emphasize @mdboom's position as the creator of pyodide in some way. For instance with the possibility to veto  decisions, if you think this would be preferable?  

Please comment if you have other ideas or suggestions.

Rendered version: https://github.com/rth/pyodide/blob/governance/docs/project/governance.md

cc @mdboom @wlach @dalcde @phorward @bcolloran

### CODE REVIEW

File: docs/project/governance.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+### Contributors
+
+Contributors are community members who contribute in concrete ways to the
+project. Anyone can become a contributor, and contributions can take many forms
+– not only code – as detailed in the {ref}`how_to_contibute`.
+
+### Community members team
+
+The community members team is composed of community members who have permission on
+Github to label and close issues. Their work is
```
Comment (rth) [inline docs/project/governance.md:27]: Not sure about the name for this team. iodode has "community members team", the other name could be "triage team".

Another side effect from being in this team is that it would be possible to request review on PRs from its members via Github UI. That feature doesn't work for users not part of the org.

File: docs/project/governance.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+* **Minor Documentation changes and minor build setup changes**, include for
+  instance improving the wording in the documentation, updating CI or
+  dependencies.  Core developers are expected to give “reasonable time” to
+  others to give their opinion on the pull request if they’re not confident
+  others would agree. If no review is received within this time, the Pull
+  Request can be merged.
+
+* **Code changes impacting user facing APIs, or having backward compatibility
+  implications** require +1 by a core developers, no -1 by a core developer
```
Comment (rth) [inline docs/project/governance.md:92]: What constitutes a backward incompatible change in pyodide is still a bit unclear. In the sense that updating emsdk could also break some backward compatibility. I imagine this means not breaking use user code that's working for some version of pyodide, excluding potential breakage due to python packages updates.

Comment (hoodmane) [reply docs/project/governance.md:92]: I'm not really sure what the text here means. Does it mean that for changes to behavior / user facing API, at least one core dev other than the one making the PR should review the PR and approve it before merging? This seems to match the current procedure?

File: docs/project/governance.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
+# pyodide governance and decision-making
+
+The purpose of this document is to formalize the governance process used by the
+pyodide project, to clarify how decisions are made and how the various
+elements of our community interact.
```
Comment (hoodmane) [inline docs/project/governance.md:5]:
```suggestion
members of our community interact.
```

File: docs/project/governance.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+This document establishes a decision-making structure that takes into account
+feedback from all members of the community and strives to find consensus, while
+avoiding any deadlocks.
+
+Anyone with an
+interest in the project can join the community, contribute to the project
+design and participate in the decision making process. This document describes
+how that participation takes place and how to set about earning merit within
+the project community.
```
Comment (hoodmane) [inline docs/project/governance.md:14]:
```suggestion
how to participate and earn merit in the pyodide community.
```

File: docs/project/governance.md (hunk 2 above)
Comment (rth) [reply docs/project/governance.md:92]: Yes, that's what it mean and  what we are mostly doing now. I'll replace "+1" by "a review and an approval (+1 vote)" and "-1" by "objection with a -1 vote" if it helps.

File: docs/project/governance.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+The community members team is composed of community members who have permission on
+Github to label and close issues. Their work is
+crucial to improve the communication in the project.
+
+Similarly to what has been decided in the [CPython project](
+https://devguide.python.org/triaging/#becoming-a-member-of-the-python-triage-team),
+any contributor may become a member of this team, after
+showing some continuity in participating in pyodide
+development (with pull requests and reviews).
```
Comment (hoodmane) [inline docs/project/governance.md:34]:
```suggestion
After participating in pyodide development with pull requests and reviews for a period of time, any contributor may become a member of the team.
The process for adding team members is modeled on the [CPython project](
https://devguide.python.org/triaging/#becoming-a-member-of-the-python-triage-team).
```

File: docs/project/governance.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  instance improving the wording in the documentation, updating CI or
+  dependencies.  Core developers are expected to give “reasonable time” to
+  others to give their opinion on the pull request if they’re not confident
+  others would agree. If no review is received within this time, the Pull
+  Request can be merged.
+
+* **Code changes impacting user facing APIs, or having backward compatibility
+  implications** require a review and an approval (+1 vote) by a core
+  developer, no objections with -1 vote by a core developer (lazy consensus).
+  This process happens on the pull-request page.
```
Comment (hoodmane) [inline docs/project/governance.md:94]:
```suggestion
  implications** require review and approval by a core
  developer and no objections raised by any core developer (lazy consensus).
  This process happens on the pull-request page.
```
Maybe say again that there should be reasonable time given and also suggest @notifying community members that one expects to have a stake in the matter? Maybe also clarify that it needs to be reviewed by a core developer different from the one who opened the PR?

I think it's confusing to say +1 vote / -1 vote because it's not clear to me that there is any formal way to do that / any formal difference between "I object to this" and "-1"?

File: docs/project/governance.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+pyodide uses a "consensus seeking" process for making decisions. The group
+tries to find a resolution that has no open objections among core developers.
+At any point during the discussion, any core-developer can call for a vote,
+which will conclude two weeks from the call for the vote. This is what we
+hereafter may refer to as “the decision making process”.
+
+Decisions (in addition to adding core developers as above)
+are made according to the following rules:
+
+* **Minor Documentation changes and minor build setup changes**, include for
```
Comment (hoodmane) [inline docs/project/governance.md:84]: Maybe these could be called "maintenance changes"

File: docs/project/governance.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+place on the project contributors' [issue
+tracker](https://github.com/pyodide/pyodide/issues) and on [Github
+discussion](https://github.com/pyodide/pyodide/discussions).
+Occasionally, sensitive discussion occurs on a private communication channels.
+
+pyodide uses a "consensus seeking" process for making decisions. The group
+tries to find a resolution that has no open objections among core developers.
+At any point during the discussion, any core-developer can call for a vote,
+which will conclude two weeks from the call for the vote. This is what we
+hereafter may refer to as “the decision making process”.
```
Comment (hoodmane) [inline docs/project/governance.md:79]: It's a little unclear to me what scenarios would require a formal vote, is it just for any issue that leads to a significant amount of contention?

File: docs/project/governance.md (hunk 5 above)
Comment (hoodmane) [reply docs/project/governance.md:34]: Do we really want to focus on PRs and reviews here? Seems like we should also emphasize helping users in various ways (answering questions on gitter, responding to issues explaining how to do stuff).

File: docs/project/governance.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+how that participation takes place and how to set about earning merit within
+the project community.
+
+## Roles And Responsibilities
+
+### Contributors
+
+Contributors are community members who contribute in concrete ways to the
+project. Anyone can become a contributor, and contributions can take many forms
+– not only code – as detailed in the {ref}`how_to_contibute`.
```
Comment (hoodmane) [inline docs/project/governance.md:22]: Would be nice to explicitly mention e.g., answering user questions as a way to contribute here.

File: docs/project/governance.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+the continued development of the project through ongoing engagement with the
+community. They have shown they can be trusted to maintain pyodide with
+care. Being a core developer allows contributors to more easily carry on
+with their project related activities by giving them direct access to the
+project’s repository and is represented as being a member of the core team on the
+pyodide [GitHub organization](https://github.com/orgs/pyodide/teams/core/members).
+Core developers are expected to review code
+contributions, can merge approved pull requests, can cast votes for and against
+merging a pull-request, and can be involved in deciding major changes to the
+API.
```
Comment (hoodmane) [inline docs/project/governance.md:51]: Might be nice to clarify the wording here, but I'm not quite clear how. Everyone can be involved in the consensus process? We explicitly want to include users who haven't contributed code in discussions about major API changes. If a lot of users don't like it, we might consider not doing it. I guess here it's intended to mean that the final up-or-down judgment calls in cases of contention go to core devs?

File: docs/project/governance.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+New core developers can be nominated by any existing core developers. Once they
+have been nominated, there will be a vote by the current core developers.
+Voting on new core developers is one of the few activities that takes place on
+the project's private communication channels. While it is expected that most votes
```
Comment (hoodmane) [inline docs/project/governance.md:56]: > private communication channels 

Is there a plan to set up private communication channels? I assume we don't currently have these.

**Review by hoodmane — COMMENTED** [review]
This sort of writing is surprisingly hard work.

File: docs/project/governance.md (hunk 1 above)
Comment (hoodmane) [reply docs/project/governance.md:27]: I think "community members team" sounds friendlier, "triage team" sounds gory and like it requires medical training.

File: docs/project/governance.md (hunk 9 above)
Comment (rth) [reply docs/project/governance.md:21]: Absolutely, but I think different ways of how to contribute, such as answering user questions, reviewing PRs should be listed in https://pyodide.readthedocs.io/en/latest/development/contributing.html not in the governance. That page needs re-organizing anyway.

For instance an example of such section in https://scikit-learn.org/stable/developers/contributing.html#ways-to-contribute

File: docs/project/governance.md (hunk 11 above)
Comment (rth) [reply docs/project/governance.md:53]: Yes, we need to decide which ones. One possibility is ask for creation of a pyodide@python.org mailing list. The other is to use the comment system specific to Github teams, but that would tend to get lost in all the other GH notifications I think.

File: docs/project/governance.md (hunk 9 above)
Comment (hoodmane) [reply docs/project/governance.md:21]: I think giving a single example like "answering user questions" here might help indicate that we mean "contribute" in a very inclusive sense.

File: docs/project/governance.md (hunk 6 above)
Comment (rth) [reply docs/project/governance.md]: > suggest notifying community members that one expects to have a stake in the matter?

I think it's up to the contributor to decide about mentioning or not. Generally maintainers are often following the repo and will see the PR happen. Whether they will respond or not depends on their availability, interest in the particular topic etc and so systematically mentioning is nor really useful if you follow all issues anyway.

> Maybe also clarify that it needs to be reviewed by a core developer different from the one who opened the PR?

You need an approval to merge those PR and you can't approve your own PRs in the Github UI so I think that excludes this case.

Comment (hoodmane) [reply docs/project/governance.md]: > you can't approve your own PRs in the Github UI

Oh okay, I hadn't tried that yet.

File: docs/project/governance.md (hunk 8 above)
Comment (rth) [reply docs/project/governance.md:76]: The only scenario where a formal note needs to happen in the current document is for adding new maintainers. All other points should ideally reach a consensus by discussion, and only fall back on voting if agreement cannot be reached and a decision needs to be taken.

**Comment by rth** [comment]
Thanks for the feedback @hoodmane! I think I addressed your comments.

File: docs/project/governance.md [hunk 12]
```diff
@@ -78,17 +79,17 @@ hereafter may refer to as “the decision making process”.
...
   dependencies.  Core developers are expected to give “reasonable time” to
   others to give their opinion on the pull request if they’re not confident
   others would agree. If no review is received within this time, the Pull
   Request can be merged.
 
-* **Code changes impacting user facing APIs, or having backward compatibility
-  implications** require review and approval by a core
-  developer and no objections raised by any core developer (lazy consensus).
-  This process happens on the pull-request page.
+* **Code changes in general, and expecially those impacting user facing APIs**,
```
Comment (hoodmane) [inline docs/project/governance.md:89]: Spelling: "expecially" ==> "especially"

File: docs/project/governance.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+contributions, can merge approved pull requests, can cast votes for and against
+merging a pull-request, and can make decisions about major changes to the
+API (all contributors are welcome to participate in the discussion).
+
+New core developers can be nominated by any existing core developers. Once they
+have been nominated, there will be a vote by the current core developers.
+Voting on new core developers is one of the few activities that takes place on
+the project's private communication channels. While it is expected that most votes
+will be unanimous, a two-thirds majority of the cast votes is enough. The vote
+needs to be open for at least 1 week.
```
Comment (phorward) [inline docs/project/governance.md:56]: "one week"?

File: docs/project/governance.md [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+Core developers that have not contributed to the project (commits or GitHub
+comments) in the past 2 years will be asked if they want to become emeritus
```
Comment (phorward) [inline docs/project/governance.md:59]: "two years"?

File: docs/project/governance.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+members of the community. All non-sensitive project management discussion takes
+place on the project contributors' [issue
+tracker](https://github.com/pyodide/pyodide/issues) and on [Github
+discussion](https://github.com/pyodide/pyodide/discussions).
+Occasionally, sensitive discussion occurs on a private communication channels.
+
+pyodide uses a "consensus seeking" process for making decisions. The group
+tries to find a resolution that has no open objections among core developers.
+At any point during the discussion, any core-developer can call for a vote,
+which will conclude two weeks from the call for the vote. This is what we
```
Comment (phorward) [inline docs/project/governance.md:76]: here, it is already "two weeks" :+1:

File: docs/project/governance.md [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+Decisions (in addition to adding core developers as above)
+are made according to the following rules:
+
+* **Maintenance changes**, include for
+  instance improving the wording in the documentation, updating CI or
+  dependencies.  Core developers are expected to give “reasonable time” to
+  others to give their opinion on the pull request if they’re not confident
+  others would agree. If no review is received within this time, the Pull
+  Request can be merged.
```
Comment (phorward) [inline docs/project/governance.md:87]: I'm not happy with these two sentences, because they're a little bit confusing.

> Core developers are expected to give “reasonable time” to
  others to give their opinion on the pull request if they’re not confident
  others would agree.

Might this be changed into `Core developers are expected to give “reasonable time” to
  others to give their opinion on the Pull Request in case they’re not confident
  that others would agree.` ?

> If no review is received within this time, the Pull Request can be merged.

Change this into `If no further review or blocking comment on the Pull Request is received within this time, it be merged.`?

**Review by phorward — APPROVED** [review]
Hello @rth, thanks for setting this up. I'm mostly happy with it and found only few sentences that could be changed.

Important note, but this is not part of the document: When setting up the new pyodide repository in the pyodide organisation, can you please only allow for "Squash merge" into master? It can be configured on Github, and this is not the case here. It would be more clear to the core developers who merge Pull Requests.

**Comment by phorward** [comment]
@rth Perfect, thanks! :+1:

**Review by mdboom — APPROVED** [review]
This looks good to me.

Regarding giving me a special position for tie-breaking, I appreciate the sentiment, but realistically these days I'm probably too busy to be responsive / have enough context for these things.

**Comment by rth** [comment]
Thanks for the feedback @mdboom !

I therefore merge this governance document as approved.
